src/java/com/puppetlabs/ssl_utils/ExtensionsUtils.java

```diff
@@ -740,61 +741,92 @@ private static SubjectKeyIdentifier publicKeyToSubjectKeyIdentifier(PublicKey pu
-                    "provided.");
-        }
+        X509Certificate cert = (X509Certificate) authKeyIdMap.get("cert");
 
-        String issuer = (String) authKeyIdMap.get ("issuer_dn");
-        if (issuer == null) {
-            if (serialNumber != null) {
-                throw new IllegalArgumentException(
-                        "'issuer' not provided for auth key identifier " +
-                        "but was expected since 'serial_number' was provided");
-            }
-        }
-        else {
-            if (serialNumber == null) {
+        if (cert == null) {
+            if (pubKey == null && serialNumber == null) {
                 throw new IllegalArgumentException(
-                        "'serial_number' not provided for auth key identifier" +
-                        "but was expected since 'issuer' was provided");
+                        "Neither 'public_key', 'serial_number', or 'cert' provided for " +
+                        "auth key identifier.  At least one of these must be " +
+                        "provided.");
             }
 
-            GeneralNames issuerAsGeneralNames =
-                    new GeneralNames(new GeneralName(new X500Name(issuer)));
-            if (authorityKeyId != null) {
-                authorityKeyId = new AuthorityKeyIdentifier(
-                        authorityKeyId.getKeyIdentifier(),
-                        issuerAsGeneralNames,
-                        serialNumber);
+            String issuer = (String) authKeyIdMap.get ("issuer_dn");
+            if (issuer == null) {
+                if (serialNumber != null) {
+                    throw new IllegalArgumentException(
+                            "'issuer' not provided for auth key identifier " +
+                            "but was expected since 'serial_number' was provided");
+                }
             }
             else {
-                authorityKeyId = new AuthorityKeyIdentifier(
-                        issuerAsGeneralNames,
-                        serialNumber);
+                if (serialNumber == null) {
+                    throw new IllegalArgumentException(
+                            "'serial_number' not provided for auth key identifier" +
+                            "but was expected since 'issuer' was provided");
+                }
+
+                GeneralNames issuerAsGeneralNames =
+                        new GeneralNames(new GeneralName(new X500Name(issuer)));
+                if (authorityKeyId != null) {
+                    authorityKeyId = new AuthorityKeyIdentifier(
+                            authorityKeyId.getKeyIdentifier(),
+                            issuerAsGeneralNames,
+                            serialNumber);
+                }
+                else {
+                    authorityKeyId = new AuthorityKeyIdentifier(
+                            issuerAsGeneralNames,
+                            serialNumber);
+                }
             }
         }
+        else {
```

Comment:
There's definitely some cleanup I could do with the repeated `DigestCalculators`, but really I've only added one nesting `if`. I would say that the diff represented here on github  makes the change way harder to parse than just looking at the file...but I have been looking at this a long time 😂

Reply:
yeah, I switched to split mode to review this because the diff is particularly hard to read compared to the actual code. I still think that the original method was borderline too big/nested and this change pushes it over the edge.